pkg/ottl/ottlfuncs/func_replace_all_matches.go

```diff
@@ -43,9 +45,22 @@ func replaceAllMatches[K any](target ottl.PMapGetter[K], pattern string, replace
 		if err != nil {
 			return nil, err
 		}
-		replacementVal, err := replacement.Get(ctx, tCtx)
-		if err != nil {
-			return nil, err
+		if fn.IsEmpty() {
+			replacementVal, err = replacement.Get(ctx, tCtx)
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			fnVal := fn.Get()
+			replacementExpr, errNew := fnVal.Get(&FuncArgs[K]{Input: replacement})
+			if errNew != nil {
+				return nil, errNew
+			}
+			replacementValRaw, errNew := replacementExpr.Eval(ctx, tCtx)
+			if errNew != nil {
+				return nil, errNew
+			}
+			replacementVal = replacementValRaw.(string)
```

Comment:
I think it would be good to check that `replacementValRaw` really is a string and error otherwise. The function uses a `StringGetter` to get `replacement`, which means it will error if `replacement` is not a string.  After the string value goes through the function, if the output is not a string we should also error.

Maybe in the future there is some type checking we can do to enforce that a function supplied in the optional arg returns a string.

Reply:
Pushed a change that now returns an error if `replacementValRaw` isn't a string.